src/Type.zig

```diff
@@ -4060,10 +4060,18 @@ pub fn getUnionLayout(loaded_union: InternPool.LoadedUnionType, zcu: *const Zcu)
 /// Handles const-ness and address spaces in particular.
 /// This code is duplicated in `Sema.analyzePtrArithmetic`.
 /// May perform type resolution and return a transitive `error.AnalysisFail`.
-pub fn elemPtrType(ptr_ty: Type, offset: ?usize, pt: Zcu.PerThread) !Type {
+/// If the ptr_ty is array_ptr type, then set is_array_ptr to true.
+pub fn elemPtrType(ptr_ty: Type, offset: ?usize, pt: Zcu.PerThread, is_array_ptr: bool) !Type {
     const zcu = pt.zcu;
     const ptr_info = ptr_ty.ptrInfo(zcu);
-    const elem_ty = ptr_ty.elemType2(zcu);
+    const elem_ty_tmp = ptr_ty.elemType2(zcu);
+    // If the ptr_ty is array_ptr type, and the pointer is a C pointer, then
+    // the pointer is same as the one size zig pointer. So we need to get
+    // shallow element type.
+    const elem_ty = if (is_array_ptr == true and ptr_info.flags.size == .c)
```

Comment:
Why do you need to add a new parameter instead of checking if `elem_ty_tmp` is an array?